Safe Transmute RFC (#5)
rust-lang/project-safe-transmute: rfcs/0000-safe-transmute.md

```rust
use core::transmute::{
    TransmuteInto,
    stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
};

#[derive(PromiseTransmutableInto, PromiseTransmutableFrom)] // declare `Foo` to be *stably* transmutable
#[repr(C)]
pub struct Foo(pub u8, pub u16);
//                    ^ there's a padding byte here, between these fields

// Transmute fearlessly!
let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!
//                  ^^^^^^^^^^^^^^ provided by the `TransmuteInto` trait

let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!

// error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
//   --> src/demo.rs:15:27
//    |
// 15 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
//    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
//    |
//   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`

> //   = note: byte 8 of the source type may be uninitialized; byte 8 of the destination type cannot be uninitialized.

Reviewer: Shouldn't this say "bit 8" instead of "byte 8" ?

# Safe Transmute RFC

- Feature Name: `safe_transmute`
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> We propose traits, namely `TransmuteInto` and `TransmuteFrom`, that are implemented *automatically* for combinations of types that may be safely transmuted. In other words, this RFC makes safe transmutation *as easy as 1..., 2..., `repr(C)`!*

Author: @rylev https://github.com/rust-lang/project-safe-transmute/pull/5/commits/a5122b2cd80d86b4c7cd7b552519fa5fe8ebb8a1 adds additional explanatory comments to this example, including one that calls attention to `TransmuteInto`

I'm going to mark this as resolved, but feel free to un-resolve if need be.

And that's not all: Kelly misses that the very layout of `Foo` itself is undefined⁠—it must be annotated with `#[repr(C)]`!

Kelley concludes:
> In Zig the problem of alignment is solved completely; the compiler catches all possible alignment issues. In the situation where you need to assert to the compiler that something is more aligned than Zig thinks it is, you can use `@alignCast`. This inserts a cheap safety check in debug mode to make sure the alignment assertion is correct.

This situation is frustrating. Like `zig`, `rustc` *knows* the layout of `Foo`, it *knows* that the layout of `Foo` is unspecified because it lacks the `#[repr(C)]` attribute, and it *knows* the minimum alignment requirements of `Foo` are stricter than that of `u8`. Why doesn't it also help us avoid UB?

This RFC proposes mechanisms that use this information to enable safer transmutation. We revisit Kelley's motivating example with these proposed mechanisms near the end of this RFC, [*here*][case-study-safer-unsafe].

## Efficient Network Parsing

> ***TODO: Josh?***

Reviewer 2: Got this on my TODO list for tomorrow :)

> ***TODO: Josh?***

Author: @joshlf still owes us a paragraph of prose, but we chatted and observed that with a slice casting extension, writing methods like `packet`'s [`take_obj_front`](https://fuchsia-docs.firebaseapp.com/rust/packet/trait.BufferView.html#method.take_obj_front) becomes trivial.

Concretely, once you have slice casting, you can provide a method like this:
```rust
impl<'a> &'a [u8] {
    /// Read a value of type `T` from the slice, if it is sufficiently long.
    fn read<T>(&'a mut self) -> Option<&'a T>
    where
        Self: CastInto<&'a [T]>,
    {
        let (parsable, remainder) = self.split_at(size_of::<T>);
        *self = remainder;
        parsable.cast_into().first()
    }
}
```

...which then lets you write stuff like this:
```rust
pub struct UdpPacket<'a> {
    hdr: &'a UdpHeader,
    body: &'a [u8],
}

#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(C)]
struct UdpHeader {
    src_port: [u8; 2],
    dst_port: [u8; 2],
    length:   [u8; 2],
    checksum: [u8; 2],
}

impl<'a> UdpPacket<'a> {
    pub fn parse(mut bytes: &[u8]) -> Option<UdpPacket> {
        Some(UdpPacket { hdr: bytes.read()?, body: bytes })
    }
}
```

> ***TODO: Josh?***

Author: Thanks to @joshlf, this is now resolved: https://github.com/rust-lang/project-safe-transmute/pull/5/commits/b7fef2d10f067c3813851408b17e16e783159f5c

```rust
/* ⚠️ This example intentionally does not compile. */
let mut x = Constrained::new(0, 0);
{
    let y : &mut u16 = (&mut x).transmute_into(); // Compile Error!
    *y = 0xFF00u16;
}

let z : Constrained = x;
```
If this example did not produce a compile error, the value of `z` would not be a safe instance of its type, `Constrained`.

### 📖 When is a transmutation stable?
[stability]: #When-is-a-transmutation-stable

Since the soundness and safety of a transmutation is affected by the layouts of the source and destination types, changes to those types' layouts may cause code which previously compiled to produce errors. In other words, transmutation causes a type's layout to become part of that type's API for the purposes of SemVer stability.

The question is, then: *how can the author of a type reason about transmutations they did not write, from-or-to types they did not write?* We address this problem by introducing two traits which both allow an author to opt-in to stability guarantees for their types, and allow third-parties to reason at compile-time about what guarantees are provided for such types.

> #### `PromiseTransmutableFrom` and `PromiseTransmutableInto`

Author: I'm fond of these names because `Promise` reflects that the user implements these traits to *make* a promise, and `Transmutable{From,Into}` corresponds exactly to what they're promising. I'm quite open to bikeshedding, though!

#### Stability & Transmutation
A `Src` type is *stably* transmutable into a `Dst` type *only if* `<Src as PromiseTransmutableInto>::Archetype` is transmutable, stability notwithstanding, into `<Dst as PromiseTransmutableFrom>::Archetype`; formally:
```rust
unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
where
    Src: PromiseTransmutableInto,
    Dst: PromiseTransmutableFrom,
    <Dst as PromiseTransmutableFrom>::Archetype:
        TransmuteFrom<
            <Src as PromiseTransmutableInto>::Archetype,
            NeglectStability
        >
{}
```

Why is this the case?

The type `<Src as PromiseTransmutableInto>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Src` that could affect its use as a source type in transmutations. Conversely, `<Dst as PromiseTransmutableFrom>::Archetype` exemplifies the furthest extreme of non-breaking changes that could be made to the layout of `Dst` that could affect its use as a destination type in transmutations. If a transmutation between these extremities is valid, then so is `Src: TransmuteInto<Dst>`.

> #### Common Use-Case: Promising a Fixed Layout

Author: @rylev I'm unsure how useful this shorthand would be.

For end-users, I anticipate that `PromiseTransmutableFrom` and `PromiseTransmutableInto` will virtually never appear in users' code outside of the context of implementing those traits. In writing the examples for this RFC, I haven't yet encountered an instance where I needed to use either of those traits in a `where` bound. So, I don't think this shorthand would be useful outside of the context of implementing the traits.

And, _within_ the context of implementing those traits, nearly all users will be using the shorthand `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]`. So here, `#[derive(FixedLayout)]` would be a shorthand for a shorthand.

Is there a use-case you have in mind?

> #### Common Use-Case: Promising a Fixed Layout

Reviewer 3: Short-hand for short-hand is not necessarily a bad thing 😄 . I believe a vast majority of the time that users will want `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]`. Given the long names and the frequency of its use `#[derive(FixedLayout)]` strikes me as an ergonomic choice. But I realize that it's not super elegant to introduce a new trait just as a short-hand for deriving.

> #### Common Use-Case: Promising a Fixed Layout

Reviewer 4: No, that's a better trait name for sure. People will understand it more when looking at docs and source.

> #### Common Use-Case: Promising a Fixed Layout

Author: @rylev  Defining a trait is rather tricky (since those derives are a little more nuanced than just `Archetype = Self` to account for field stability declarations), but it's very technically feasible to provide a `derive(PromiseTransmutableFromAndInto)` macro that would be equivalent to `derive(PromiseTransmutableFrom, PromiseTransmutableInto)` — *without* a `PromiseTransmutableFromAndInto` trait.

> #### Common Use-Case: Promising a Fixed Layout

Author: > Defining a trait is rather tricky (since those derives are a little more nuanced than just Archetype = Self to account for field stability declarations)

To clarify a little more about why having a `FixedLayout` shorthand/trait is tricky, it's because the concept of "fixed layout" is a little tricky.

---

First, there's the name: `PromiseTransmutableFrom` isn't *really* making a promise about layout, it's making a promise about future transmutability. Transmutability involves factors beyond layout, like [constructability](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#-when-is-a-transmutation-safe). If we wanted to create a `derive` shorthand for `derive(PromiseTransmutableFrom, PromiseTransmutableInto)`, the name `PromiseFixedLayout` (or `PromiseStableLayout`, etc...) misrepresents what's actually being promised.

---

Second, there's the semantics: what does promising a fixed layout *mean*? You're probably mean to promise that *you* won't make changes to the type's transmutability, but what about *others*? What if your type's fields come from other crates? What if your type's fields are generic?

The behavior of `#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]` reflects this inherent complexity: it **doesn't** mean your type has *totally-assured-transmutability*, it means your type has *as-assured-as-possible-transmutability*. Concretely: When you write `#[derive(PromiseTransmutable{From,Into})]`, your type [inherits the instability of its fields](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#common-use-case-promising-a-fixed-layout). If its fields have strong stability promises, so too will your type; if its fields have weak stability promises, so too will your type.

---

To conclude, I don't think there's a good mapping of a [mem-markers](https://github.com/rust-secure-code/mem-markers)-style `FixedLayout` abstraction onto  `PromiseTransmutableFrom` and `PromiseTransmutableInto`. While it might seem simpler, at first glance, to just present *one* trait (`FixedLayout`) instead of two (`PromiseTransmutableFrom` and `PromiseTransmutableInto`), the concept of a `FixedLayout` isn't well-defined, and the complexities of `PromiseTransmutable{From,Into}` reflect the complexity *inherent* to the problem of transmutation stability.

> #### Common Use-Case: Promising a Fixed Layout

Reviewer 3: What I'm arguing here is that there is a use case for "POD" that will be quite common. By "POD" in this case I mean types where the in memory-representation of that type is intrinsic to its functionality and changing that would be a breaking change. This is common in the parsing scenario for instance. Having some sort of short hand for saying this type's layout/alignment/etc cannot change would be a simple "happy path" for many use cases. I'm not arguing this as a replacement to `PromiseTransmutable{From,Into}` but as a compliment. This doesn't have to be a `derive` either. I can imagine also `#[repr(C, fixed)]` for instance.

> #### Common Use-Case: Promising a Fixed Layout

Author: Per a conversation with @rylev, https://github.com/rust-lang/project-safe-transmute/pull/5/commits/ed172fe84874c154118f2054435ceb8ff94312e1 proposes a `#[derive(PromiseTransmutable)]` ergonomic extension. I'm fairly [convinced](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#extension-promisetransmutable-shorthand) that adding such a shorthand is the right thing to do and I [refer to it in the Guide-Level Explanation](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#common-use-case-as-stable-as-possible), but it's formally defined as an Extension because it's technically non-central to the RFC and poses a few unusual design quirks.

To promise that type's layout is completely fixed, simply `derive` the `PromiseTransmutableFrom` and `PromiseTransmutableInto` traits:
```rust
#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);
```
This expands to:
```rust
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);

// Generated stability declarations for `Foo`:

> const _: () = {

Reviewer 5: How can 2 custom derives result in only 1 top-level item?

Author: Whoops, you're right. This *should* result in two separate items.

Author: Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/7a883702a5390f8419689dfc188f263a665ac871

    #[repr(C)]

>     struct TransmutableFromArchetype

Reviewer 5: ```suggestion
    struct TransmutableFromArchetype(
```

>     struct TransmutableFromArchetype

Author: Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/7a883702a5390f8419689dfc188f263a665ac871.

        <Bar as PromiseTransmutableFrom>::Archetype,
        <Baz as PromiseTransmutableFrom>::Archetype,
    );

    #[repr(C)]

>     struct TransmutableIntoArchetype

Reviewer 5: ```suggestion
    struct TransmutableIntoArchetype(
```

> // Generated stability declarations for `Foo`:
> const _: () = {
> 
>     #[repr(C)]
>     struct TransmutableFromArchetype(
>         <Bar as PromiseTransmutableFrom>::Archetype,
>         <Baz as PromiseTransmutableFrom>::Archetype,
>     );
> 
>     #[repr(C)]
>     struct TransmutableIntoArchetype(
>         <Bar as PromiseTransmutableInto>::Archetype,
>         <Baz as PromiseTransmutableInto>::Archetype,
>     );
> 
>     impl PromiseTransmutableFrom for Foo {
>         type Archetype = TransmutableFromArchetype;
>     }
> 
>     impl PromiseTransmutableInto for FooTransmutableIntoArchetype {
>         type Archetype = TransmutableIntoArchetype;
>     }
> };

Author: This span is [incorrect](https://github.com/rust-lang/project-safe-transmute/pull/5/files#r464604585) and *should* be:
```suggestion
/// Generated `PromiseTransmutableFrom` for `Foo`
const _: () = {
    use core::transmute::stability::PromiseTransmutableFrom;

    #[repr(C)]
    pub struct TransmutableFromArchetype(
        pub <Bar as PromiseTransmutableFrom>::Archetype,
        pub <Baz as PromiseTransmutableFrom>::Archetype,
    );

    impl PromiseTransmutableFrom for Foo {
        type Archetype = TransmutableFromArchetype;
    }
};

/// Generated `PromiseTransmutableInto` for `Foo`
const _: () = {
    use core::transmute::stability::PromiseTransmutableInto;

    #[repr(C)]
    pub struct TransmutableIntoArchetype(
        pub <Bar as PromiseTransmutableInto>::Archetype,
        pub <Baz as PromiseTransmutableInto>::Archetype,
    );

    impl PromiseTransmutableInto for Foo {
        type Archetype = TransmutableIntoArchetype;
    }
};
```

            mem::forget(src);
            dst
        }
    }
}

// implemented in terms of `TransmuteFrom`
pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
where
    Neglect: UnsafeTransmuteOptions,
{
    fn transmute_into(self) -> Dst
    where
        Self: Sized,
        Dst: Sized,
        Neglect: SafeTransmuteOptions;

    unsafe fn unsafe_transmute_into(self) -> Dst
    where
        Self: Sized,
        Dst: Sized,
        Neglect: UnsafeTransmuteOptions;
}

unsafe impl<Src, Dst, Neglect> TransmuteInto<Dst, Neglect> for Src
where
    Src: ?Sized,
    Dst: ?Sized + TransmuteFrom<Src, Neglect>,
    Neglect: UnsafeTransmuteOptions,
{
    ...
}
```

In the above definitions, `Src` represents the source type of the transmutation, `Dst` represents the destination type of the transmutation, and `Neglect` is a parameter that [encodes][options] which static checks the compiler ought to neglect when considering if a transmutation is valid. The default value of `Neglect` is `()`, which reflects that, by default, the compiler does not neglect *any* static checks.

### Neglecting Static Checks
[options]: #Neglecting-Static-Checks

> The default value of the `Neglect` parameter, `()`, statically forbids transmutes that are unsafe, unsafe, or unstable. However, you may explicitly opt-out of some static checks:

Reviewer 5: ```suggestion
The default value of the `Neglect` parameter, `()`, statically forbids transmutes that are unsafe, unsound, or unstable. However, you may explicitly opt-out of some static checks:
```

> The default value of the `Neglect` parameter, `()`, statically forbids transmutes that are unsafe, unsafe, or unstable. However, you may explicitly opt-out of some static checks:

Author: Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/651e3bfe2181f56709d2ccfc13346b73bfb89aac.

| Transmute Option    | Compromises | Usable With                                             |
|---------------------|-------------|---------------------------------------------------------|
| `NeglectStabilty`   | Stability   | `transmute_{from,into}`, `unsafe_transmute_{from,into}` |
| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |

> | `NeglectValidity`   | Soudness    | `unsafe_transmute_{from,into}`                          |

Reviewer 5: ```suggestion
| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |
```

# Summary
[summary]: #summary

We propose traits, namely `TransmuteInto` and `TransmuteFrom`, that are implemented *automatically* for combinations of types that may be safely transmuted. In other words, this RFC makes safe transmutation *as easy as 1..., 2..., `repr(C)`!*
```rust
use core::transmute::{
    TransmuteInto,
    stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
};

#[derive(PromiseTransmutableInto, PromiseTransmutableFrom)]
#[repr(C)]
pub struct Foo(pub u8, pub u16);
//                   ^ there's a padding byte here

// Transmute fearlessly!
let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!

let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!

// error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
//   --> src/demo.rs:14:27
//    |
// 14 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
//    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
//    |
//   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`
```

# Motivation
[motivation]: #motivation

> Beyond the titular goal of providing a mechanism of safe transmutation, this RFC seeks to provide mechanisms which:
>  - [authoritatively codify language layout guarantees](#codifying-language-layout-guarantees)
>  - [allow crate authors to codify their types' layout stability guarantees](#expressing-library-layout-guarantees)
>  - [allow crate authors to codify their abstractions' layout requirements](#expressing-layout-requirements)
> 
> We foresee a range of practical consequences of fulfilling these goals, including [making unsafe Rust safer](#making-unsafe-rust-safer), providing [safer initialization primitives](#safer-initialization-primitives), and [generic atomics](#atomict). We explore these possibilities, and more, in this section.

Author: ```suggestion
We foresee a range of practical consequences of fulfilling these goals, including [making unsafe Rust safer](#making-unsafe-rust-safer), providing [safer initialization primitives](#safer-initialization-primitives), and [generic atomics](#atomict). We explore these use-cases, and more, in this section.

A *comprehensive* approach to safe transmutation provides benefits beyond the mere act of transmutation; namely:
 - [authoritatively codifies language layout guarantees](#codifying-language-layout-guarantees)
 - [allows crate authors to codify their types' layout stability guarantees](#expressing-library-layout-guarantees)
 - [allows crate authors to codify their abstractions' layout requirements](#expressing-layout-requirements)
```

For instance, a common bit-packing technique involves abusing the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.

The mechanisms proposed by the RFC enable this. We return to this example near the end of the RFC, [here][case-study-alignment].

## Making Unsafe Rust Safer
[motivation-safer-unsafe]: #Making-Unsafe-Rust-Safer
In the blog post [*Unsafe Zig is Safer than Unsafe Rust*](https://andrewkelley.me/post/unsafe-zig-safer-than-unsafe-rust.html), Andrew Kelley asks readers to consider the following Rust code:

```rust
struct Foo {
    a: i32,
    b: i32,
}

fn main() {
    unsafe {
        let mut array: [u8; 1024] = [1; 1024];
        let foo = mem::transmute::<&mut u8, &mut Foo>(&mut array[0]);
        foo.a += 1;
    }
}
```

> This pattern, Kelly notes, is common when interacting with operating system APIs.

Reviewer 6: nit: Kelley

This rust program compiles, but quietly introduces undefined behavior. `Foo` requires stricter in-memory alignment than `u8`. The equivalent code in Zig produces a compiler error for this reason and requires modification.

> And that's not all: Kelly misses that the very layout of `Foo` itself is undefined⁠—it must be annotated with `#[repr(C)]`!

Reviewer 6: nit: Kelley

Kelley concludes:
> In Zig the problem of alignment is solved completely; the compiler catches all possible alignment issues. In the situation where you need to assert to the compiler that something is more aligned than Zig thinks it is, you can use `@alignCast`. This inserts a cheap safety check in debug mode to make sure the alignment assertion is correct.

This situation is frustrating. Like `zig`, `rustc` *knows* the layout of `Foo`, it *knows* that the layout of `Foo` is unspecified because it lacks the `#[repr(C)]` attribute, and it *knows* the minimum alignment requirements of `Foo` are stricter than that of `u8`. Why doesn't it also help us avoid UB?

This RFC proposes mechanisms that use this information to enable safer transmutation. We revisit Kelley's motivating example with these proposed mechanisms near the end of this RFC, [*here*][case-study-safer-unsafe].

## Efficient Network Parsing
***TODO: Josh?***

## Safer Initialization Primitives

> The initialization primitives `mem::zeroed<T>` and `mem::uninitialized<T>` are `unsafe` because they may be used to initialized types for which zeroed or uninitialized memory are *not* valid bit-patterns. The `mem::zeroed` function recently gained a dynamically-enforced validity check, but this safety measure isn't wholly satisfactory: the validity properties of `T` are statically known, but the check is dynamic.

Reviewer 6: nit: Should be "initialize types" and not "initialized types"

struct Padded(pub u8, pub u16, pub u8);

#[repr(C)]
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Packed(pub u16, pub u16, pub u16);

assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
```

We may safely transmute from `Packed` to `Padded`:
```rust
let _ : Padded = Packed::default().transmute_into();
```
...but not from `Padded` to `Packed`:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : Packed = Padded::default().transmute_into(); // Compile Error!
```
...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.

##### Preserve or Shrink Size
[owned-size]: #Preserve-or-Shrink-Size

It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
```rust
let _ : [u8; 16] = [0u8; 32]::default().transmute_into();
```
This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.

A value may ***not*** be transmuted into a type of greater size:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : [u8; 32] = [0u8; 16]::default().transmute_into(); // Compile Error!
```

> 

Reviewer 6: This may be obvious, but this doesn't necessarily make it clear *why* it can't be transmuted into a type of greater size. The previous example makes it clear that doing the transmute would expose two uninitialized padding bytes. I presume the same problem is true in this case (the extra bytes would be uninitialized) but it may be worth calling that out for clarity.

Author: > I presume the same problem is true in this case (the extra bytes would be uninitialized) but it may be worth calling that out for clarity.

This is an interesting observation, because I hadn't considered that expanding the size could *ever* be valid, except your observation suggests a case where it could be: transmuting `[MaybeUninit<u8>; 2]` to `[MaybeUninit<u8>; 4]` (or any situation where the trailing bit of the destination type may be uninit)!

I'm going to rephrase this as:
> A value may not be transmuted into a type of greater size, **if doing so would expose uninitialized bytes as if they were initialized**:

Author: Resolved by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/1a119b21bc3ca7c447c1bd645cb0805d0c2b5b3c

#### Implicit Constructability
A struct or enum variant is *fully implicitly constructable* at a given location only if, at that location, that type can be instantiated via its *implicit constructor*, and its fields are also *implicitly constructable*.

The *implicit constructor* of a struct or enum variant is the constructor Rust creates implicitly from its definition; e.g.:
```rust
struct Point<T> {
    x: T,
    y: T,
}

let p = Point { x: 4, y: 2 };
     // ^^^^^^^^^^^^^^^^^^^^ An instance of `Point` is created here, via its implicit constructor.
```

Limiting implicit constructability is the fundamental mechanism with which type authors build safe abstractions for `unsafe` code, whose soundness is dependent on preserving invariants on fields. Usually, this takes the form of restricting the visibility of fields. For instance, consider the type `NonEmptySlice`, which enforces a validity constraint on its fields via its constructor:

```rust

> pub mod crate_a {
> 
>     #[repr(C)]
>     pub struct NonEmptySlice<T> {
>         data: *const T,
>         len: usize,
>     }
> 
>     impl<T> NonEmptySlice<T> {
>         pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
>             assert!(len <= N);
>             assert!(len > 0);
>             Self {
>                 data: arr as *const T,
>                 len,
>             }
>         }
> 
>         pub fn first(&self) -> &T {
>             unsafe { &*self.data }
>         }
>     }
> 
> }

Author: ```suggestion
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<'a, T> {
        data: *const T,
        len: usize,
        lifetime: core::marker::PhantomData<&'a ()>,
    }

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N);
            assert!(len > 0);
            Self {
                data: arr as *const T,
                len,
                lifetime: core::marker::PhantomData,
            }
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.data }
        }
    }

}
```

```
It is sound for `first` to be a *safe* method is because the `from_array` constructor ensures that `data` is safe to dereference, and because `from_array` is the *only* way to safely initialize `NonEmptySlice` outside of `crate_a` (note that `NonEmptySlice`'s fields are *not* `pub`). As a rule: any field that is not marked `pub` should be assumed to be private *because* it is subject to safety invariants.

Unfortunately, field visibility modifiers are not a surefire indicator of whether a type is *fully* implicitly constructable. A type author may restrict the implicit constructability of a type even in situations where all fields of that type (*and all fields of those fields*) are `pub`; consider:
```rust

> pub mod crate_a {
> 
>     #[repr(C)]
>     pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
> 
>     impl<T> NonEmptySlice<T> {
>         pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
>             assert!(len <= N && len > 0);
>             Self(
>                 private::NonEmptySliceInner {
>                     data: arr as *const T,
>                     len,
>                 }
>             )
>         }
> 
>         pub fn first(&self) -> &T {
>             unsafe { &*self.0.data }
>         }
>     }
> 
>     // introduce a private module to avoid `private_in_public` error (E0446):
>     pub(crate) mod private {
>         #[repr(C)]
>         pub struct NonEmptySliceInner<T> {
>             pub data: *const T,
>             pub len: usize,
>         }
>     }
> 
> }

Author: ```suggestion
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```

```
In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.

#### Constructability and Transmutation
Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author.

> In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)

Author: ```suggestion
In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```

However, we recognize that this definition poses implementation challenges: In our definition of constructability, answering the question of whether a struct or enum variant is constructible depends on *where* that question is being asked. Consequently, answering whether a given type `Src` is `TransmutableInto` a given type `Dst` will depend on *where* that question is posed.

We recommend adopting a simplified definition of *constructability*: a type is *constructible* if its fields are marked `pub`, and those fields are constructible. With this definition, answering the question of whether a type is constructible does *not* depend on where the question is asked: we do not examine the visibility of the involved types.

Unfortunately, with no other actions taken, this simplified definition comes...

##### ...at the Cost of Safety
This definition is *usually* sufficient for ensuring safety: it is *generally* an error to expose a private type in a public type signature. However, these errors may be circumvented using the public-type-in-private-module trick:
```rust

> pub mod crate_a {
> 
>     #[repr(C)]
>     pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
> 
>     impl<T> NonEmptySlice<T> {
>         pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
>             assert!(len <= N && len > 0);
>             Self(
>                 private::NonEmptySliceInner {
>                     data: arr as *const T,
>                     len,
>                 }
>             )
>         }
> 
>         pub fn first(&self) -> &T {
>             unsafe { &*self.0.data }
>         }
>     }
> 
>     // introduce a private module to avoid `private_in_public` error (E0446):
>     pub(crate) mod private {
>         #[repr(C)]
>         pub struct NonEmptySliceInner<T> {
>             pub data: *const T,
>             pub len: usize,
>         }
>     }
> 
> }

Author: ```suggestion
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```

```
With this simplified definition of constructability, it is possible for a third-party to define a *safe* constructor of `NonEmptySlice` that produces a value which is *unsafe* to use:
```rust

> pub evil_constructor<T>(src: T) -> NonEmptySlice<u8>
> where
>     T: TransmuteInto<NonEmptySlice<u8>, NeglectStability>,

Author: ```suggestion
pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
where
    T: TransmuteInto<NonEmptySlice<'static, u8>, NeglectStability>,
```

{
    src.transmute_into()
}

evil_constructor(0u128).first() // muahaha!
```

The above code is "safe" because our simplified definition of constructability fails to recognize this pattern of encapsulation, and because `NeglectStability` is a `SafeTransmutationOption`.

The intent of `NeglectStability` is to permit the safe transmutation of types that predate the stabilization of the stability declaration traits. It also provides a convenient escape-hatch for type authors to neglect the stability of transmutations of their *own* types, without sacrificing safety. `NeglectStability` is a `SafeTransmutationOption` because, in principle, neglecting stability does not diminish safety. Our simplified definition of constructability violates this principle.

By temporarily sacrificing these goals, we may preserve safety solely...

##### ...at the Cost of `NeglectStability`
We may preserve safety by demoting `NeglectStability` to `UnsafeTransmutationOption`-status.

In doing so, a third-party is forced to resort to an `unsafe` transmutation to construct `NonEmptySlice`; e.g.:

```rust

> pub evil_constructor<T>(src: T) -> NonEmptySlice<u8>
> where
>     T: TransmuteInto<NonEmptySlice<u8>, NeglectStability>,

Author: ```suggestion
pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
where
    T: TransmuteInto<NonEmptySlice<'static, u8>, NeglectStability>,
```

{
    // unsafe because we `NeglectStability`
    unsafe { src.unsafe_transmute_into() }
}
```

Demoting `NeglectStability` to unsafe-status does not stop type authors from opting-in to stable (and thus safe) transmutations; e.g., with `derive(PromiseTransmutableFrom)`:
```rust

> pub mod crate_a {
> 
>     #[derive(PromiseTransmutableFrom)]
>     #[repr(C)]
>     pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
> 
>     impl<T> NonEmptySlice<T> {
>         pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
>             assert!(len <= N && len > 0);
>             Self(
>                 private::NonEmptySliceInner {
>                     data: arr as *const T,
>                     len,
>                 }
>             )
>         }
> 
>         pub fn first(&self) -> &T {
>             unsafe { &*self.0.data }
>         }
>     }
> 
>     // introduce a private module to avoid `private_in_public` error (E0446):
>     pub(crate) mod private {
>         #[derive(PromiseTransmutableFrom)]
>         #[repr(C)]
>         pub struct NonEmptySliceInner<T> {
>             pub data: *const T,
>             pub len: usize,
>         }
>     }
> 
> }

Author: ```suggestion
pub mod crate_a {

    #[derive(PromiseTransmutableFrom)]
    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[derive(PromiseTransmutableFrom)]
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```

```
In the above example, the type author declares `NonEmptySlice` and `NonEmptySliceInner` to be stably instantiatable via transmutation. Given this, a third-party no longer needs to resort to `unsafe` code to violate the the invariants on `inner`:
```rust

> pub evil_constructor<T>(src: T) -> NonEmptySlice<u8>
> where
>     T: TransmuteInto<NonEmptySlice<u8>>,

Author: ```suggestion
pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
where
    T: TransmuteInto<NonEmptySlice<'static, u8>>,
```

        + PromiseTransmutableInto<Sum<Neglect, NeglectStability>>;
}
```
Implementations of these traits for a given `Neglect` declares that a transmutation which is accepted while neglecting a particular set of checks (namely the set encoded by `Neglect`) will *continue* to be possible.

We omit these definition from this RFC's recommendations because they are not completely satisfying. For instance, `Neglect` is a *logically* unordered set of options, but is encoded as a tuple (which *is* ordered). To declare a transmutation that requires neglecting validity and alignment checks as stable, only *one* of these impls ought to be necessary:

```rust
impl PromiseTransmutableFrom<(NeglectAlignment, NeglectValidity)> for Foo
{
    ...
}

impl PromiseTransmutableFrom<(NeglectValidity, NeglectAlignment)> for Foo
{
    ...
}
```
Writing *both* impls (as we do above) is logically nonsense, but is nonetheless supported by Rust's coherence rules.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

## Case Studies

### Case Study: Making Unsafe Rust Safer
[case-study-safer-unsafe]: #Case-Study-Making-Unsafe-Rust-Safer

> This RFC proposes mechanisms for safe transmutation and safer *unsafe* transmutation. How might [Kelly's motivating example][motivation-safer-unsafe] change in light of these mechanisms?

Author: ```suggestion
This RFC proposes mechanisms for safe transmutation and safer *unsafe* transmutation. How might [Kelley's motivating example][motivation-safer-unsafe] change in light of these mechanisms?
```

# Safe Transmute RFC

- Feature Name: `safe_transmute`
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

We propose traits, namely `TransmuteInto` and `TransmuteFrom`, that are implemented *automatically* for combinations of types that may be safely transmuted. In other words, this RFC makes safe transmutation *as easy as 1..., 2..., `repr(C)`!*
```rust

> use core::transmute::{
>     TransmuteInto,
>     stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
> };
> 
> #[derive(PromiseTransmutableInto, PromiseTransmutableFrom)]
> #[repr(C)]
> pub struct Foo(pub u8, pub u16);
> //                   ^ there's a padding byte here
> 
> // Transmute fearlessly!
> let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!
> 
> let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
> 
> // error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
> //   --> src/demo.rs:14:27
> //    |
> // 14 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
> //    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
> //    |
> //   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`

Author: Attempt to improve https://github.com/rust-lang/project-safe-transmute/pull/5#discussion_r464500451
```suggestion
use core::transmute::{
    TransmuteInto,
    stability::{PromiseTransmutableInto, PromiseTransmutableFrom},
};

#[derive(PromiseTransmutableInto, PromiseTransmutableFrom)] // declare `Foo` to be *stably* transmutable
#[repr(C)]
pub struct Foo(pub u8, pub u16);
//                    ^ there's a padding byte here, between these fields

// Transmute fearlessly!
let _ : Foo = 64u32.transmute_into(); // Alchemy Achieved!
//                  ^^^^^^^^^^^^^^ provided by the `TransmuteInto` trait

let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!

// error[E0277]: the trait bound `u32: TransmuteFrom<foo::Foo, _>` is not satisfied
//   --> src/demo.rs:15:27
//    |
// 15 | let _ : u32 = Foo(16, 12).transmute_into(); // Compile Error!
//    |                           ^^^^^^^^^^^^^^ the trait `TransmuteFrom<foo::Foo, _>` is not implemented for `u32`
//    |
//   = note: required because of the requirements on the impl of `TransmuteInto<u32, _>` for `foo::Foo`
```

```rust
#[repr(C)]
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Padded(pub u8, pub u16, pub u8);

#[repr(C)]
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Packed(pub u16, pub u16, pub u16);

assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
```

We may safely transmute from `Packed` to `Padded`:
```rust
let _ : Padded = Packed::default().transmute_into();
```
...but not from `Padded` to `Packed`:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : Packed = Padded::default().transmute_into(); // Compile Error!
```
...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.

##### Preserve or Shrink Size
[owned-size]: #Preserve-or-Shrink-Size

It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
```rust
let _ : [u8; 16] = [0u8; 32]::default().transmute_into();
```
This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.

> A value may ***not*** be transmuted into a type of greater size:

Author: ```suggestion
A value may ***not*** be transmuted into a type of greater size, if doing so would expose uninitialized bytes as initialized:
```

```rust
#[derive(PromiseTransmutableFrom)]
#[repr(align(8)]
struct Aligned8;
```
and then using this `where` bound:
```rust
where
    &T: TransmuteInto<&Aligned8>
```

## Rationale: `TransmuteFrom`/`TransmuteInto`

### Why support arbitrary transmutation?
Some [prior art][prior-art], especially in the crate ecosystem, provides an API that [only supports transmutations involving particular types](#Source-and-Destination-Types-Supported) (e.g., from/into bytes). As we discuss in the [prior art][prior-art] section, we believe that the inflexibility of such approaches make them a poor basis of a language proposal. In particular, these restrictive approaches don't leave room to grow: supporting additional transmutations requires additional traits.

The API advocated by this proposal is unopinionated about what transmutations users might wish to do, and what transmutations the compiler is able to reason about. The implementation of this RFC may be initially very simple (and perhaps support no more than the restrictive approaches allow for), but then subsequently grow in sophistication—*without* necessitating public API changes.

### Why *two* traits?

> If `TransmuteInto` is implemented in terms of `TransmuteFrom`, why provide it at all? We do so for consistency with libcore's [`From`/`Into`](https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html) traits, and because directionality conveys intent: `TransmuteInto` connotes *conversion*, whereas `TransmuteInto` connotes initialization. We believe that the supporting code examples of this RFC demonstrate the explanatory benefits of providing *both* traits.

Reviewer 6: I believe this should read:

`TransmuteInto` connotes *conversion*, whereas `TransmuteFrom` connotes initialization

With the second one being `TransmuteFrom`.

Author: ```suggestion
If `TransmuteInto` is implemented in terms of `TransmuteFrom`, why provide it at all? We do so for consistency with libcore's [`From`/`Into`](https://doc.rust-lang.org/stable/rust-by-example/conversion/from_into.html) traits, and because directionality conveys intent: `TransmuteFrom` connotes *conversion*, whereas `TransmuteInto` connotes initialization. We believe that the supporting code examples of this RFC demonstrate the explanatory benefits of providing *both* traits.
```

#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Packed(pub u16, pub u16, pub u16);

assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
```

We may safely transmute from `Packed` to `Padded`:
```rust
let _ : Padded = Packed::default().transmute_into();
```
...but not from `Padded` to `Packed`:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : Packed = Padded::default().transmute_into(); // Compile Error!
```
...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.

##### Preserve or Shrink Size
[owned-size]: #Preserve-or-Shrink-Size

It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
```rust
let _ : [u8; 16] = [0u8; 32]::default().transmute_into();
```
This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.

A value may ***not*** be transmuted into a type of greater size:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : [u8; 32] = [0u8; 16]::default().transmute_into(); // Compile Error!
```

#### Requirements on References
[transmute-references]: #transmuting-references

> The [restrictions above that apply to transmuting owned values][transmute-owned] also apply to transmuting references. However, references carry a few additional restrictions.
> 
> A [sound transmutation] must:
>  - [preserve or relax alignment][reference-alignment],
>  - [preserve or shrink lifetimes][reference-lifetimes],
>  - [preserve or shrink uniqueness][reference-mutability], and
>  - and if the destination type is a mutate-able reference, [preserve validity][reference-validity].
> 

Author: The no-size-expansion rule [*could*](https://github.com/rust-lang/project-safe-transmute/pull/5#discussion_r465307607) be relaxed for value transmutations, but it cannot be relaxed for reference transmutations (since that 'extra' space may be memory already used for something else). If that rule is relaxed for value transmutations, it needs to be re-added to reference transmutations:

````suggestion
The [restrictions above that apply to transmuting owned values][transmute-owned] also apply to transmuting references. However, references carry a few additional restrictions.

A [sound transmutation] must:
 - [preserve or shrink size][reference-size],
 - [preserve or relax alignment][reference-alignment],
 - [preserve or shrink lifetimes][reference-lifetimes],
 - [preserve or shrink uniqueness][reference-mutability], and
 - and if the destination type is a mutate-able reference, [preserve validity][reference-validity].

##### Preserve or Shrink Size
[reference-size]: #Preserve-or-Shrink-Size

You may preserve or decrease the size of the referent type via transmutation:
```rust
let _: &[u8; 3] = (&[0u8; 9]).transmute_into();
```

However, you may **not**, under any circumstances, *increase* the size of the referent type:
```rust
/* ⚠️ This example intentionally does not compile. */
let _: &[u8; 9] = (&[0u8; 3]).transmute_into(); // Compile Error! 
```
````

Our proposed API supports complex, granular stability declarations, all-the-while retaining simplicity in simple use-cases (namely, one can simply the stability declaration traits). 

## Alternative: Implementing this RFC in a Crate

This RFC builds on ample [prior art][prior-art] in the crate ecosystem, but these efforts strain against the fundamental limitations of crates. Fundamentally, safe transmutation efforts use traits to expose layout information to the type system. The burden of ensuring safety [is usually either placed entirely on the end-user, or assumed by complex, incomplete proc-macro `derives`][mechanism-manual].

> An exception to this rule is the [typic][crate-typic] crate, which complex, type-level programming to emulate a compiler-supported, "smart" `TransmuteFrom` trait (like the one proposed in this RFC). Nonetheless, [typic][crate-typic] is fundamentally limited: since Rust does not provide a type-level mechanism for reflecting over the structure of arbitrary types, even [typic][crate-typic] cannot judge the safety of a transmutation without special user-added annotations on type definitions. Although [typic][crate-typic] succeeds as a proof-of-concept, its maintainability is questionable, and the error messages it produces are [lovecraftian](https://en.wikipedia.org/wiki/Lovecraftian_horror).

Reviewer 6: Missing word: Perhaps, 'which *utilizes* complex, type-level programming'?

Author: ```suggestion
An exception to this rule is the [typic][crate-typic] crate, which utilizes complex, type-level programming to emulate a compiler-supported, "smart" `TransmuteFrom` trait (like the one proposed in this RFC). Nonetheless, [typic][crate-typic] is fundamentally limited: since Rust does not provide a type-level mechanism for reflecting over the structure of arbitrary types, even [typic][crate-typic] cannot judge the safety of a transmutation without special user-added annotations on type definitions. Although [typic][crate-typic] succeeds as a proof-of-concept, its maintainability is questionable, and the error messages it produces are [lovecraftian](https://en.wikipedia.org/wiki/Lovecraftian_horror).
```

```
...because not all instances of [`u8`] are valid instances of `Bool`.

Another example: While laying out certain types, Rust may insert padding bytes between the layouts of fields. In the below example `Padded` has two padding bytes, while `Packed` has none:
```rust
#[repr(C)]
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Padded(pub u8, pub u16, pub u8);

#[repr(C)]
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
struct Packed(pub u16, pub u16, pub u16);

assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
```

We may safely transmute from `Packed` to `Padded`:
```rust
let _ : Padded = Packed::default().transmute_into();
```
...but not from `Padded` to `Packed`:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : Packed = Padded::default().transmute_into(); // Compile Error!
```
...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.

##### Preserve or Shrink Size
[owned-size]: #Preserve-or-Shrink-Size

It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
```rust

> let _ : [u8; 16] = [0u8; 32]::default().transmute_into();

Author: ```suggestion
let _ : [u8; 16] = [u8; 32]::default().transmute_into();
```

```
This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.

A value may ***not*** be transmuted into a type of greater size, if doing so would expose uninitialized bytes as initialized:
```rust
/* ⚠️ This example intentionally does not compile. */

> let _ : [u8; 32] = [0u8; 16]::default().transmute_into(); // Compile Error!

Author: ```suggestion
let _ : [u8; 32] = [u8; 16]::default().transmute_into(); // Compile Error!
```

#### Common Use-Case: Promising a Fixed Layout
To promise that type's layout is completely fixed, simply `derive` the `PromiseTransmutableFrom` and `PromiseTransmutableInto` traits:
```rust
#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);
```
This expands to:
```rust
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);

/// Generated `PromiseTransmutableFrom` for `Foo`
const _: () = {
    use core::transmute::stability::PromiseTransmutableFrom;

    #[repr(C)]
    pub struct TransmutableFromArchetype(
        pub <Bar as PromiseTransmutableFrom>::Archetype,
        pub <Baz as PromiseTransmutableFrom>::Archetype,
    );

    impl PromiseTransmutableFrom for Foo {
        type Archetype = TransmutableFromArchetype;
    }
};

/// Generated `PromiseTransmutableInto` for `Foo`
const _: () = {
    use core::transmute::stability::PromiseTransmutableInto;

    #[repr(C)]
    pub struct TransmutableIntoArchetype(
        pub <Bar as PromiseTransmutableInto>::Archetype,
        pub <Baz as PromiseTransmutableInto>::Archetype,
    );

    impl PromiseTransmutableInto for Foo {
        type Archetype = TransmutableIntoArchetype;
    }
};
```

> #### Uncommon Use Case: Complex Stability Guarantees

Author: ```suggestion
#### Uncommon Use-Case: Weak Stability Guarantees
[stability-uncommon]: #uncommon-use-case-weak-stability-guarantees
```

```rust
pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
where
    T: TransmuteInto<NonEmptySlice<'static, u8>>,
{
    src.transmute_into()
}

evil_constructor(0u128).first() // muahaha!
```
This safety hazard is not materially different from the one that would be induced if the type author implemented `DerefMut<Target=NonEmptySliceInner>` for `NonEmptySlice`, or made the `private` module `pub`, or otherwise explicitly provided outsiders with unrestricted mutable access to `data`.

##### Recommendation
We recommend that that implementers of this RFC initially simplify constructability by:
 - adopting our simplified definition of constructability
 - demoting `NeglectStability` to unsafe status (i.e., not implementing `SafeTransmuteOptions` for `NeglectStability`; *only* `UnsafeTransmuteOptions`)

If and when the implementation of `TransmuteFrom` encodes our complete definition of constructability, `NeglectStability` shall become a safe transmute option.

## Dissecting Stability

> In this section, we iteratively review the design of the stability declaration traits, starting with a simplest-possible design:
> ```rust
> pub unsafe trait PromiseTransmutableFrom
> {
>     type Archetype;
> }
> 
> pub unsafe trait PromiseTransmutableInto
> {
>     type Archetype;
> }
> ```

Author: Clarify the purpose of *Dissecting Stability*—it's about turning user-facing complexity into complexity the type system handles.

````suggestion
In this section, we iteratively review the design of the stability declaration traits, starting with a simplest-possible design:
```rust
pub unsafe trait PromiseTransmutableFrom
{
    type Archetype;
}

pub unsafe trait PromiseTransmutableInto
{
    type Archetype;
}
```
This section gives concrete examples of how misusing these simple, unsafe definitions can lead to unsoundness, and then refines them until they are immune to misuse (and, consequently, *safe*).
````

        + PromiseTransmutableFrom;
}

pub trait PromiseTransmutableInto
{
    type Archetype
        : TransmuteFrom<Self, NeglectStability>
        + PromiseTransmutableInto;
}
```

# Drawbacks
[drawbacks]: #drawbacks

## No Notion of Platform Stability
The stability declaration traits communicate library layout stability, but not *platform* layout stability. A transmutation is platform-stable if it compiling one one platform implies it will compile on all other platforms. Unfortunately, platform-unstable types are common;  e.g.:

- All pointer-related primitive types (`usize`, `isize`, `*const T`, `*mut T`, `&T`, `&mut T`) possess platform-dependent layouts; their sizes and alignments are well-defined, but vary between platforms. Concretely, whether `usize` is `TransmuteInto<[u8; 4]>` or `TransmuteInto<[u8; 8]>` will depend on  the platform.
- The very existence of some types depends on platform, too; e.g., the contents of [`core::arch`](https://doc.rust-lang.org/stable/core/arch/), [`std::os`](https://doc.rust-lang.org/stable/std/os/), and [`core::sync::atomic`](https://doc.rust-lang.org/stable/std/sync/atomic/) all depend on platform.

> Expanding this RFC to address such types would introduce considerable additional complexity:

Author: ```suggestion
Our proposed stability system is oblivious to the inter-platform variations of these types. Expanding our stability system to be aware of inter-platform variations would introduce considerable additional complexity:
```

While it is unclear if there is any demand for this degree of flexibility, this upgrade-path should be carefully considered *before* stabilizing (and thus committing) to this RFC's layout stability declaration traits.

### Questions Out of Scope
We consider the following unresolved questions to be out-of-scope of *this* RFC process:

##### Design of `NeglectConstructability`?
`TransmuteFrom` and `TransmuteInto` require that the destination type has a matching constructor in which all fields are marked `pub`. Conspicuously *missing* from this RFC is a `NeglectConstructability` unsafe option to disable this check.

The omission is intentional. The consequences of such an option are suprising in both their subtlety and their unsafety. Some of unsafe Rust's hairiest interactions lie at the intersections of `!Send`, `!Sync`, `UnsafeCell` and restricted field visibility. These building blocks are used to build safe, public abstractions that encapsulate unsafe, hidden internals.

# Future possibilities
[future-possibilities]: #future-possibilities

> 
> 

Author: ````suggestion

## Extension: `PromiseTransmutable` Shorthand
[extension-promisetransmutable-shorthand]: #extension-promisetransmutable-shorthand

We anticipate that *most* users will merely want to [promise][stability-common] that their types are as-stable-as-possible. To do so, this RFC proposes the shorthand:
```rust
#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);
```
As a shorthand, this is still rather long. For such users, the separability of `PromiseTransmutableFrom` and `PromiseTransmutableInto` is totally irrelevant. We therefore propose a `derive(PromiseTransmutable)` shorthand, such that this:
```rust
#[derive(PromiseTransmutable)]
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);
```
...is equivalent to this:
```rust
#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(C)]
pub struct Foo(pub Bar, pub Baz);
```

However, we caution *against* adding a corresponding trait or trait alias; e.g.:
```rust
trait PromiseTransmutable = PromiseTransmutableFrom + PromiseTransmutableInto;
```
The vast majority of users will *only* confront the stability declaration traits in the context of deriving them; the *only* scenario in which end-users will refer to these traits in a type-context is the [rare use-case][stability-uncommon] of *manually* implementing them. For such users, the separability of `PromiseTransmutableFrom` and `PromiseTransmutableInto` *is* relevant. The availability of a `PromiseTransmutable` trait or trait alias in this scenario would be a distraction, since referring to it in a type-context is almost certainly a misstep.

We acknowledge that it is unusual for a `derive` macro to not create an item of the same name, but this weirdness is outweighed by the weirdness of the alternative: providing a trait for which there is almost no good use. 

````

## Extension: `include_data!`
[future-possibility-include_data]: #Extension-include_data

With the mechanisms proposed in this RFC, we may define a structured-data alternative to `include_bytes`:
```rust
macro_rules! include_data {
    ($file:expr) => {{
        use core::transmute::TransmuteInto;
        include_bytes!($file).transmute_into()
    }};
}
```
...and, with this, replace all instances of `unsafe` in our [earlier][motivation-including-data] motivating example:
```rust
pub fn recognize(input: &Matrix<f64, U1, U784>) -> usize
{
    static WEIGHT: &Matrix<f64, U784, U10> = include_data!("/weight.bin");

    static BIAS: &Matrix<f64, U1, U10> = include_data!("/bias.bin")

    network::recognize(input, WEIGHT, BIAS)
}
```

## Extension: Reference Casting
[ext-ref-casting]: #Extension-Reference-Casting
```rust
/// Try to convert a `&T` into `&U`.
///
/// This produces `None` if the referent isn't appropriately
/// aligned, as required by the destination type.
pub fn try_cast_ref<'t, 'u, T, U>(src: &'t T) -> Option<&'u U>
where

>     &'t T: UnsafeTransmuteInto<&'u U, neglect::Alignment>,

Author: Update to the current name for this option.
```suggestion
    &'t T: UnsafeTransmuteInto<&'u U, NeglectAlignment>,
```

This RFC proposes mechanisms that programmers will use to confidently answer such questions—by checking whether the `TransmuteFrom` and `TransmuteInto` traits are implemented, or (equivalently) by checking whether the `can_transmute` predicate (a `const fn`) is satisfied.

## Expressing Library Layout Guarantees
There is no canonical way for crate authors to declare the SemVer layout guarantees of their types. Crate authors currently must state their layout guarantees using prose in their documentation. In contrast to structural stability (e.g., the declared visibility of fields), layout stability is expressed extra-linguistically.

This isn't satisfactory: guarantees expressed in prose outside of the Rust programming language are guarantees that cannot be reasoned about *inside* the language. Whereas `rustc` can dutifully deny programmers access to private fields, it is unable to prevent programmers from making unfounded expectations of types' in-memory layouts.

> This RFC proposes simple-but-powerful [mechanisms][stability] for declaring layout stability guarantees.

Reviewer 6: Broken link: stability

For instance, a common bit-packing technique involves abusing the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.

The mechanisms proposed by the RFC enable this. We return to this example near the end of the RFC, [here][case-study-alignment].

## Making Unsafe Rust Safer
[motivation-safer-unsafe]: #Making-Unsafe-Rust-Safer
In the blog post [*Unsafe Zig is Safer than Unsafe Rust*](https://andrewkelley.me/post/unsafe-zig-safer-than-unsafe-rust.html), Andrew Kelley asks readers to consider the following Rust code:

```rust
struct Foo {
    a: i32,
    b: i32,
}

fn main() {
    unsafe {
        let mut array: [u8; 1024] = [1; 1024];
        let foo = mem::transmute::<&mut u8, &mut Foo>(&mut array[0]);
        foo.a += 1;
    }
}
```
This pattern, Kelley notes, is common when interacting with operating system APIs.

> This rust program compiles, but quietly introduces undefined behavior. `Foo` requires stricter in-memory alignment than `u8`. The equivalent code in Zig produces a compiler error for this reason and requires modification.

Reviewer 6: ```suggestion
This Rust program compiles, but quietly introduces undefined behavior. `Foo` requires stricter in-memory alignment than `u8`. The equivalent code in Zig produces a compiler error for this reason and requires modification.
```

This situation is frustrating. Like `zig`, `rustc` *knows* the layout of `Foo`, it *knows* that the layout of `Foo` is unspecified because it lacks the `#[repr(C)]` attribute, and it *knows* the minimum alignment requirements of `Foo` are stricter than that of `u8`. Why doesn't it also help us avoid UB?

This RFC proposes mechanisms that use this information to enable safer transmutation. We revisit Kelley's motivating example with these proposed mechanisms near the end of this RFC, [*here*][case-study-safer-unsafe].

## Efficient Network Parsing
***TODO: Josh?***

## Safer Initialization Primitives
The initialization primitives `mem::zeroed<T>` and `mem::uninitialized<T>` are `unsafe` because they may be used to initialize types for which zeroed or uninitialized memory are *not* valid bit-patterns. The `mem::zeroed` function recently gained a dynamically-enforced validity check, but this safety measure isn't wholly satisfactory: the validity properties of `T` are statically known, but the check is dynamic.

> This RFC proposes mechanisms that would allow these functions to explicitly bound their generic parameter `T` based on its validity properties. We revisit this motivating use-case with our proposed mechanisms near the end of this RFC, [*here*][future-possibility-safe-initialization].

Reviewer 6: Broken link: future-possibility-safe-initialization

pub AtomicTrilean(AtomicU8);

impl AtomicTrilean {

    pub const fn new(v: Trilean) -> Self {
        AtomicTrilean(
          AtomicU8::new(
            unsafe { mem::transmute(v) }
          ))
    }

    ...
}
```

The mechanisms proposed by this RFC would eliminate this pattern, permitting truly generic `Atomic` types. We revisit this motivating use-case with these proposed mechanisms near the end of this RFC, [*here*][future-possibility-generic-atomics].

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Terminology & Concepts

### 📖 Transmutation
**Transmutation** is the act of reinterpreting the bytes corresponding to a value of one type as if they corresponded to a different type.

### 📖 Soundness
A transmutation is ***sound*** if the mere act of transmuting a value from one type to another is not compiler undefined behavior.

### 📖 Safety
A sound transmutation is ***safe*** if *using* the transmuted value cannot violate memory safety.

### 📖 Stability
A safe transmutation is ***stable*** if the authors of the source type and destination types have indicated that the layouts of those types is part of their libraries' stability guarantees.

> ## Concepts in Depth

Author: Full disclosure: If this RFC were accepted and fully implemented tonight, some of the examples in this section still wouldn't quite work *as-written* due to an existing inference issue; e.g.:
```rust
let _ : i8 = i8::transmute_from(0u8);
//              ^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteFrom`
let _ : i8 = 0u8.transmute_into();
//               ^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteInto`
```
https://github.com/rust-lang/rust/issues/27336 might be related to what's going on here. In these situations, this syntax may be used instead:
```rust
let _ : i8 = TransmuteFrom::<_>::transmute_from(0u8);
let _ : i8 = TransmuteInto::<_>::transmute_into(0u8);
```
Given that solution, I'm surprised that *this* **doesn't** work:
```rust
let _ : i8 = TransmuteFrom::transmute_from(0u8);
//           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteFrom`
let _ : i8 = TransmuteInto::transmute_into(0u8);
//           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error[E0282]: cannot infer type for type parameter `Neglect` declared on the trait `TransmuteInto`
```
...because I would think that `Trait::method` ought to be equivalent to `Trait::<_>::method`.

---

I don't know *if* or *where* to file this issue in the RFC. It's not a drawback of the RFC per se, and it's possible the issue will be resolved by the time this RFC is stabilized.

In contrast, most `struct` and `enum` types defined without an explicit `#[repr(C)]` attribute do ***not*** have well-specified layout characteristics.

To ensure that types you've define are soundly transmutable, you almost always (with very few exceptions) must mark them with the `#[repr(C)]` attribute.

#### Requirements on Owned Values
[transmute-owned]: #transmuting-owned-values

Transmutations involving owned values must adhere to two rules to be sound. They must:
 * [preserve or broaden the bit validity][owned-validity], and
 * [preserve or shrink the size][owned-size].

##### Preserve or Broaden Bit Validity
[owned-validity]: #Preserve-or-Broaden-Bit-Validity
[`NonZeroU8`]: https://doc.rust-lang.org/beta/core/num/struct.NonZeroU8.html

The bits of any valid instance of the source type must be a bit-valid instance of the destination type.

For example, we are permitted to transmute a `Bool` into a [`u8`]:
```rust
#[derive(Default, PromiseTransmutableFrom, PromiseTransmutableInto)]
#[repr(u8)]
enum Bool {
    True = 1,
    False = 0,
}

let _ : u8 = Bool::True.transmute_into();
let _ : u8 = Bool::False.transmute_into();
```
<sup>

> (Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#When-is-a-transmutation-stable?">library stability guarantee</a>.)

Reviewer 6: Broken link: #When-is-a-transmutation-stable?

Author: ```suggestion
(Note: <code>#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]</code> annotation connotes that <i>all</i> aspects of <code>Bool</code>'s layout are part of its <a href="#-when-is-a-transmutation-stable">library stability guarantee</a>.)
```

struct Packed(pub u16, pub u16, pub u16);

assert_eq!(mem::size_of::<Packed>(), mem::size_of::<Padded>());
```

We may safely transmute from `Packed` to `Padded`:
```rust
let _ : Padded = Packed::default().transmute_into();
```
...but not from `Padded` to `Packed`:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : Packed = Padded::default().transmute_into(); // Compile Error!
```
...because doing so would expose two uninitialized padding bytes in `Padded` as if they were initialized bytes in `Packed`.

##### Preserve or Shrink Size
[owned-size]: #Preserve-or-Shrink-Size

It's completely sound to transmute into a type with fewer bytes than the source type; e.g.:
```rust
let _ : [u8; 16] = [u8; 32]::default().transmute_into();
```
This transmute truncates away the final sixteen bytes of the `[u8; 32]` value.

A value may ***not*** be transmuted into a type of greater size, if doing so would expose uninitialized bytes as initialized:
```rust
/* ⚠️ This example intentionally does not compile. */
let _ : [u8; 32] = [u8; 16]::default().transmute_into(); // Compile Error!
```

#### Requirements on References
[transmute-references]: #transmuting-references

> The [restrictions above that apply to transmuting owned values][transmute-owned] also apply to transmuting references. However, references carry a few additional restrictions.

Reviewer 6: Broken link: transmute-owned

> A [sound transmutation] must:

Reviewer 6: Broken link: sound transmutation

|---------------------|-------------|---------------------------------------------------------|
| `NeglectStabilty`   | Stability   | `transmute_{from,into}`, `unsafe_transmute_{from,into}` |
| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |
| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |

`NeglectStabilty` implements the `SafeTransmuteOptions` and `UnsafeTransmuteOptions` marker traits, as it can be used in both safe and unsafe code. The selection of multiple options is encoded by grouping them as a tuple; e.g., `(NeglectAlignment, NeglectValidity)` is a selection of both the `NeglectAlignment` and `NeglectValidity` options.

We introduce two marker traits which serve to group together the options that may be used with safe transmutes, and those which may be used with `unsafe` transmutes:
```rust
pub trait SafeTransmuteOptions: private::Sealed
{}

pub trait UnsafeTransmuteOptions: SafeTransmuteOptions
{}

impl SafeTransmuteOptions for () {}
impl UnsafeTransmuteOptions for () {}
```

#### `NeglectStability`
[`NeglectStability`]: #neglectstability

> By default, `TransmuteFrom` and `TransmuteInto`'s methods require that the [layouts of the source and destination types are SemVer-stable][stability]. The `NeglectStability` option disables this requirement.

Reviewer 6: Broken link: stability

```rust
pub struct NeglectStability;

impl SafeTransmuteOptions for NeglectStability {}
impl UnsafeTransmuteOptions for NeglectStability {}
```

> Prior to the adoption of the [stability declaration traits][stability], crate authors documented the layout guarantees of their types with doc comments. The `TransmuteFrom` and `TransmuteInto` traits and methods may be used with these types by requesting that the stability check is neglected; for instance:

Reviewer 6: Broken link: stability

### Distinguishing Stability from `#[repr(C)]`
In our proposal, `#[repr(C)]` does not connote any promises of layout stability for SemVer purposes. It's [been suggested](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/RFC.3A.20Stability.20Declaration.20Traits/near/204011238) that the presence of `#[repr(C)]` *already* connotes total layout stability; if so, explicit stability declaration traits are superfluous. However, we are unaware of any documentation indicating that `#[repr(C)]` carries this implication. (It's also [been suggested](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/typic/near/201165897) that `#[repr(C)]` should *not* connote stability, because that would pose a stability hazard. We agree with this assessment.)

### Permitting Granular Stability Declarations

> The [stability declaration traits][stability] permit granular and incomplete promises of layout stability (e.g., guaranteeing the size and validity qualities of a type, but *not* its alignment. An alternative formulation of stability might be all-or-nothing (i.e., equivalent to only being able to write `Archetype = Self`). However, members of the safe-transmute working group have [expressed](https://rust-lang.zulipchat.com/#narrow/stream/216762-project-safe-transmute/topic/Transmutability.20Intrinsic/near/202712834) an interest in granular stability declarations.

Reviewer 6: Broken link: stability

An exception to this rule is the [typic][crate-typic] crate, which utilizes complex, type-level programming to emulate a compiler-supported, "smart" `TransmuteFrom` trait (like the one proposed in this RFC). Nonetheless, [typic][crate-typic] is fundamentally limited: since Rust does not provide a type-level mechanism for reflecting over the structure of arbitrary types, even [typic][crate-typic] cannot judge the safety of a transmutation without special user-added annotations on type definitions. Although [typic][crate-typic] succeeds as a proof-of-concept, its maintainability is questionable, and the error messages it produces are [lovecraftian](https://en.wikipedia.org/wiki/Lovecraftian_horror).

The development approaches like [typic][crate-typic]'s could, perhaps, be eased by stabilizing [frunk](https://crates.io/crates/frunk)-like structural reflection, or (better yet) by stabilizing a compiler plugin API for registering "smart" traits like `TransmuteFrom`. However, we suspect that such features would be drastically harder to design and stabilize. 

> Regardless of approach, almost all [prior art][prior-art] attempts to reproduce knowledge *already* possessed by rustc during the compilation process (i.e., the layout qualities of a concrete type). Emulating the process of layout computation to any degree is an error-prone duplication of effort between rustc and the crate, in a domain where correctness is crucial.

Reviewer 6: ```suggestion
Regardless of approach, almost all [prior art][prior-art] attempts to reproduce knowledge *already* possessed by `rustc` during the compilation process (i.e., the layout qualities of a concrete type). Emulating the process of layout computation to any degree is an error-prone duplication of effort between `rustc` and the crate, in a domain where correctness is crucial.
```

        Neglect: TransmuteOptions,
    {
        #[inline(always)]
        fn transmute_into(self) -> Dst
        where
            Self: Sized,
            Dst: Sized,
            Neglect: SafeTransmuteOptions,
        {
            Dst::transmute_from(self)
        }

        #[inline(always)]
        unsafe fn unsafe_transmute_into(self) -> Dst
        where
            Self: Sized,
            Dst: Sized,
            Neglect: TransmuteOptions,
        {
            unsafe { Dst::unsafe_transmute_from(self) }
        }
    }

    /// `Self: TransmuteInto<Src, Neglect`, if the compiler accepts the stability,
    /// safety, and soundness of transmuting `Src` into `Self`, notwithstanding
    /// a given set of static checks to `Neglect`.
    /* #[lang = "transmute_from"] */
    pub unsafe trait TransmuteFrom<Src: ?Sized, Neglect = ()>
    where
        Neglect: TransmuteOptions,
    {
        /// Reinterpret the bits of a value of one type as another type, safely.
        #[inline(always)]
        fn transmute_from(src: Src) -> Self
        where
            Src: Sized,
            Self: Sized,
            Neglect: SafeTransmuteOptions,
        {
            use core::{mem, ptr};

>             unsafe {
>                 let dst = ptr::read(&src as *const Src as *const Self);
>                 mem::forget(src);
>                 dst
>             }

Reviewer 7: This violates Stacked Borrows when `Src` is a type that asserts exclusivity (_e.g._, `Src = Box<T>, Dst = &'any T`).

```suggestion
            unsafe {
                let src_no_drop = ManuallyDrop::new(src);
                let dst = ptr::read(&*src_no_drop as *const Src as *const Self);
                dst
            }
```
(or `ManuallyDrop::take`, or `transmute_copy`, but these one do not generalize to the `read_unaligned` cases below)

```rust
/// A type is transmutable into itself.
unsafe impl<T, Neglect> TransmuteFrom<T, Neglect> for T
where
    Neglect: UnsafeTransmuteOptions
{}

/// A transmutation is *stable* if...
unsafe impl<Src, Dst> TransmuteFrom<Src> for Dst
where
    Src: PromiseTransmutableInto,
    Dst: PromiseTransmutableFrom,
    <Dst as PromiseTransmutableFrom>::Archetype:
        TransmuteFrom<
            <Src as PromiseTransmutableInto>::Archetype,
            NeglectStability
        >
{}
```
This [minimal implementation][minimal-impl] is sufficient for convincing the compiler to accept basic stability declarations, such as those of Rust's primitive types. It is *insufficient* for making the compiler accept transmutations between *different* types (and, consequently, complex stability declarations). Implementers should use this as a starting point.

### Listing for Initial, Minimal Implementation
[minimal-impl]: #Listing-for-Initial-Minimal-Implementation

> **[This module listing provides the minimal-viable implementation of this RFC (excepting the automatic derives).](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=25a83d8d8e33e9f631a64b53e3c39844)** This listing is also the **canonical specification** of this RFC's public API surface.

Author: ```suggestion
**[This module listing provides the minimal-viable implementation of this RFC (excepting the automatic derives).](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=bca1ad92882760e94cd4a79bccf4de30)** This listing is also the **canonical specification** of this RFC's public API surface.
```

[Simplify](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9ffa28df10593c249795ad93a7a64ebf) the minimal implementation by removing the `private` module.

This (commented-out) bound remains problematic:
```rust
    /// `Self: TransmuteInto<Dst, Neglect`, if the compiler accepts the stability,
    /// safety, and soundness of transmuting `Self` into `Dst`, notwithstanding
    /// a given set of static checks to `Neglect`.
    pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>
    where
        /*Dst: TransmuteFrom<Self, Neglect>,*/ // FIXME: Why is adding this bound so problematic?
        Neglect: UnsafeTransmuteOptions,
    {
```

Author: I've [managed](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9cd07f2ac2ecb95f558f103435554c63) to un-comment the bound, but doing so requires a lot of [new](https://gist.github.com/jswrenn/19aad724cb2e804fae6abd31a12988ff/revisions#diff-1fb9bba94efc1d84c71b0bd852912496), seemingly-redundant bounds anywhere `TransmuteInto` appears.

Author: [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html) doesn't have this reflexivity bound, so maybe it's not even a desirable thing to have?

#![feature(never_type)] // for stability declarations on `!`
#![allow(unused_unsafe, incomplete_features)]

/// Transmutation conversions.
// suggested location: `core::convert`
pub mod transmute {

    use {options::*, stability::*};

    /// Reinterprets the bits of a value of one type as another type, safely.
    #[inline(always)]
    pub /*const*/ fn safe_transmute<Src, Dst, Neglect>(src: Src) -> Dst
    where
        Src: TransmuteInto<Dst, Neglect>,
        Neglect: SafeTransmuteOptions
    {
        src.transmute_into()
    }

    /// Reinterprets the bits of a value of one type as another type, potentially unsafely.
    #[inline(always)]
    pub /*const*/ unsafe fn unsafe_transmute<Src, Dst, Neglect>(src: Src) -> Dst
    where
        Src: TransmuteInto<Dst, Neglect>,
        Neglect: UnsafeTransmuteOptions
    {
        unsafe { src.unsafe_transmute_into() }
    }

    /// `Self: TransmuteInto<Dst, Neglect`, if the compiler accepts the stability,
    /// safety, and soundness of transmuting `Self` into `Dst`, notwithstanding
    /// a given set of static checks to `Neglect`.
    pub unsafe trait TransmuteInto<Dst: ?Sized, Neglect = ()>

>     where
>         Neglect: UnsafeTransmuteOptions,

Author: **Unresolved question: *Should this additional bound be added here:***
```suggestion
    where
        Dst: TransmuteFrom<Self, Neglect>,
        Neglect: UnsafeTransmuteOptions,
```
A bound that communicates that `TransmuteInto` is the reciprocal of `TransmuteFrom` seems desirable, but introducing it [requires a lot of seemingly redundant bounds elsewhere](https://gist.github.com/jswrenn/19aad724cb2e804fae6abd31a12988ff/revisions#diff-1fb9bba94efc1d84c71b0bd852912496). The [`Into`](https://doc.rust-lang.org/std/convert/trait.Into.html) trait doesn't have this reflexivity bound, so maybe it's not actually a desirable thing to have?

</li>
<li>

**Ergonomic Complexity:** Platform instabilities are contagious: a type that *contains* a platform-unstable type is, itself, platform-unstable. Due to the sheer virulence of types with platform-dependent layouts, an explicit '`NeglectPlatformStability`' option would need to be used for *many* simple transmutations. The ergonomic cost of this would also be substantial.

</li>
<li>

**Implementation Complexity:** The mechanisms proposed by this RFC are, fundamentally, applications of and additions to Rust's type system (i.e., they're traits). Mechanisms that impact platform stability, namely `#[cfg(...)]` annotations, long precede type-resolution and layout computation in the compilation process. For instance, it's possible to define types with impossible layouts:    
```rust

> #[cfg(any())]

Reviewer 8: This is wrong. The Recursive type here is actually never instantiated. `#[cfg(any())]` is equivalent to `#[cfg(false)]`. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e39b13b30504b6c05dfd47dcdbad1603)

You may decrease the size of your type without violating stability to any size less-than-or-equal to the size of your archetype. Unfortunately, the maximum size of types in Rust is [architecture-dependent](https://github.com/rust-lang/rust/blob/63b441aafbf52d6ba789ecc478455800c1a48df9/src/librustc_target/abi/mod.rs#L179-L197). For the greatest portability, your archetype should be no larger than 2<sup>15</sup>. An archetype of this size provides (un)reasonable future flexibility.

###### Archetype Validity & Visibility
You may freely make your type *less* constructible (in terms of both theoretical validity and constructor visibility) than your archetype.

A minimally-informative archetype may be constructed using this as a building-block:
```rust
struct None(u8);
```
Since the `u8` field is private, its valid values are assumed to be constrained by library invariants.

##### Crafting Uninformative `PromiseTransmutableInto` Archetypes

###### Archetype Alignment
You may freely increase the required alignment of your type to any value greater-than-or-equal to the required alignment of your archetype. The minimum alignment requirement permitted by Rust is 1.

###### Archetype Size

> You may decrease the size of your type without violating stability to any size less-than-or-equal to the size of your archetype. The minimum size of types in Rust is 0.

Reviewer 9: Shouldn't this say that types can _increase_ their size to be greater-than-or-equal to the size of their `PromiseTransmutableInto` archetype? From the "`PromiseTransmutableFrom` and `PromiseTransmutableInto`" section above, `<Self as PromiseTransmutableInto>::Archetype` is a type that `Self` promises it can always be transmuted into; making `Self` smaller would leave bytes of the archetype uninitialized.

Author: You're right!

## Expressing Library Layout Guarantees
There is no canonical way for crate authors to declare the SemVer layout guarantees of their types. Crate authors currently must state their layout guarantees using prose in their documentation. In contrast to structural stability (e.g., the declared visibility of fields), layout stability is expressed extra-linguistically.

This isn't satisfactory: guarantees expressed in prose outside of the Rust programming language are guarantees that cannot be reasoned about *inside* the language. Whereas `rustc` can dutifully deny programmers access to private fields, it is unable to prevent programmers from making unfounded expectations of types' in-memory layouts.

This RFC proposes simple-but-powerful [mechanisms][stability] for declaring layout stability guarantees.

## Expressing Layout Requirements
Similarly, there is no canonical way for crate authors to declare the layout requirements of generic abstractions over types that have certain layout properties. 

> For instance, a common bit-packing technique involves abusing the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.

Reviewer 10: ```suggestion
For instance, a common bit-packing technique involves using the relationship between allocations and alignment. If a type is aligned to 2<sup>n</sup>, then the *n* least significant bits of pointers to that type will equal `0`. These known-zero bits can be packed with data. Since alignment cannot be currently reasoned about at the type-level, it's currently impossible to bound instantiations of a generic parameter based on minimum alignment.
```

Tiny nit, but using a property for something it may not have been intended for is still `use`, not `abuse`.

[stability-uncommon]: #uncommon-use-case-weak-stability-guarantees

If you are *most people* and want to declare your types as layout-stable, you should follow the advice in the previous sections. In doing so, you declare that you will *not* modify the layout of your type in virtually any way, except as a breaking change. If your type's fields have simple stability guarantees, this effects the strongest possible declaration of stability: it declares that *all* transmutations that are safe are *also* stable.

However, if you anticipate that you will need to make changes to your type's layouts in a non-breaking way, you can craft uninformative archetypes that effect *weaker* promises of stability. With uninformative archetypes, only a *subset* of transmutations that are safe are also considered stable. For instance, you may use `PromiseTransmutableInto` to specify an lower-bound for the size of a type, but *not* make promises about its bit-validity:

```rust
#[repr(C)] pub struct Foo(u8, u8, u8, u8, u8, u8);

#[repr(C)]
struct FooArchetype(pub [MaybeUninit<u8>; 5]);

impl PromiseTransmutableInto for Foo {
    type Type = [MaybeUninit<u8>; 5];
}
```

> ##### Crafting Uninformative `PromiseTransmutableFrom` Archetypes

Reviewer 11: I loved reading this and I'm so excited about it. For me this section was the hardest part to wrap my head around, I kept thinking that some of the language in here might be the wrong way around for what it's discussing. I still don't grok it enough to actually know if any of it is the wrong way around without spending like 30 minutes drawing diagrams. Which means my actual comment is not necessarily that there's anything wrong here but that this section might benefit from a diagram or some other wording to make this easier to follow.

Author: I've [revamped this section](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#uncommon-use-case-weak-stability-guarantees), replacing it with an example focusing in-depth on alignment!

            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```
In the above example, the type author declares `NonEmptySlice` and `NonEmptySliceInner` to be stably instantiatable via transmutation. Given this, a third-party no longer needs to resort to `unsafe` code to violate the the invariants on `inner`:
```rust
pub evil_constructor<T>(src: T) -> NonEmptySlice<'static, u8>
where
    T: TransmuteInto<NonEmptySlice<'static, u8>>,
{
    src.transmute_into()
}

evil_constructor(0u128).first() // muahaha!
```
This safety hazard is not materially different from the one that would be induced if the type author implemented `DerefMut<Target=NonEmptySliceInner>` for `NonEmptySlice`, or made the `private` module `pub`, or otherwise explicitly provided outsiders with unrestricted mutable access to `data`.

##### Recommendation
We recommend that that implementers of this RFC initially simplify constructability by:
 - adopting our simplified definition of constructability
 - demoting `NeglectStability` to unsafe status (i.e., not implementing `SafeTransmuteOptions` for `NeglectStability`; *only* `TransmuteOptions`)

>  - advise users that implementing the stability declaration traits on types that are not fully-implicitly constructable will be a compiler-error will be a compiler error (i.e., these traits must not be implemented on types exploiting the pub-in-priv trick)

Reviewer 12: ```suggestion
 - advise users that implementing the stability declaration traits on types that are not fully-implicitly constructable will be a compiler error (i.e., these traits must not be implemented on types exploiting the pub-in-priv trick)
```

"will be a compiler-error" twice.

     // ^^^^^^^^^^^^^^^^^^^^ An instance of `Point` is created here, via its implicit constructor.
```

Limiting implicit constructability is the fundamental mechanism with which type authors build safe abstractions for `unsafe` code, whose soundness is dependent on preserving invariants on fields. Usually, this takes the form of restricting the visibility of fields. For instance, consider the type `NonEmptySlice`, which enforces a validity constraint on its fields via its constructor:

```rust
pub mod crate_a {

    #[repr(C)]
    pub struct NonEmptySlice<T> {
        data: *const T,
        len: usize,
    }

    impl<T> NonEmptySlice<T> {
        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
            assert!(len <= N);
            assert!(len > 0);
            Self {

>                 data: arr as *const T,

Reviewer 3: nit: this should be -
```rust
data: arr.as_ptr(),
```

Author: [`arr as *const T` compiles fine](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=22836fa19db72d4775aabec53ccfe1b3); is this just a style nit?

Reviewer 3: Misread the code. This code isn't safe since it's possible that the reference has a shorter lifetime than the NonEmptySlice struct and thus the user can dereference the pointer after the data has been destroyed. To make it safe, you'd need to tie the lifetime of `arr` to a lifetime in `NonEmptySlice<'a, T>`

Author: Fixed by https://github.com/rust-lang/project-safe-transmute/pull/5/commits/00167d659e8c94c7f92468831714df47dadceab8

        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                }
            )
        }

        pub fn first(&self) -> &T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[repr(C)]
        pub struct NonEmptySliceInner<T> {
            pub data: *const T,
            pub len: usize,
        }
    }

}
```
In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.

#### Constructability and Transmutation
Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author.

> In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)

Reviewer 3: " would memory safety." -> " would violate memory safety."

Author: ```suggestion
In the previous examples, it would be *unsafe* to transmute `0u128` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety. (However, it's completely safe to transmute `NonEmptySlice` into a `u128`.)
```

| `NeglectAlignment`  | Safety      | `unsafe_transmute_{from,into}`                          |
| `NeglectValidity`   | Soundness   | `unsafe_transmute_{from,into}`                          |

> `NeglectStabilty` implements the `SafeTransmuteOptions` and `TransmuteOptions` marker traits, as it can be used in both safe and unsafe code. The selection of multiple options is encoded by grouping them as a tuple; e.g., `(NeglectAlignment, NeglectValidity)` is a selection of both the `NeglectAlignment` and `NeglectValidity` options.

Reviewer 3: FWIW: I think a more common naming scheme in Rust is for the default to be safe and non-default to be explicitly marked as unsafe. Would it be better for `TransmuteOptions` to be safe and for there to be `UnsafeTransmuteOptions`

}
```

> ### Removing the Dependency on Const Generics

Reviewer 3: Not sure if the extra explanation on const-generics is needed. The RFC is very long, and very detailed and think some of the information is not strictly necessary for evaluating the RFC.
